alchemiscale/compute/service.py

```diff
@@ -74,59 +78,143 @@ def __init__(
+
+        self.int_sleep = InterruptableSleep()
+        self.logger = logging.getLogger("AlchemiscaleSynchronousComputeService")
+        self.logger.setLevel(loglevel)
+
+        self.logger.addHandler(logging.StreamHandler())
+
         self._stop = False
 
-    def heartbeat(self):
+    def _register(self):
+        """Register this compute service with the compute API."""
+        self.client.register(self.compute_service_id)
+
+    def _deregister(self):
+        """Deregister this compute service with the compute API."""
+        self.client.deregister(self.compute_service_id)
+
+    def beat(self):
         """Deliver a heartbeat to the compute API, indicating this service is still alive."""
-        ...
+        self.client.heartbeat(self.compute_service_id)
+        self.logger.info("Updated heartbeat")
+
+    def heartbeat(self):
+        """Start up the heartbeat, sleeping for `self.heartbeat_interval`"""
+        while True:
+            if self._stop:
+                break
+            self.beat()
+            time.sleep(self.heartbeat_interval)
 
-    def get_tasks(self, count=1) -> List[Optional[ScopedKey]]:
+    def claim_tasks(self, count=1) -> List[Optional[ScopedKey]]:
         """Get a Task to execute from compute API.
 
         Returns `None` if no Task was available matching service configuration.
 
         """
+        # list of tasks to return
+        tasks = []
+
         taskhubs: Dict[ScopedKey, TaskHub] = self.client.query_taskhubs(
             scopes=self.scopes, return_gufe=True
         )
 
-        # based on weights, choose taskhub to draw from
-        taskhub: List[ScopedKey] = random.choices(
-            list(taskhubs.keys()), weights=[tq.weight for tq in taskhubs.values()]
-        )[0]
+        if len(taskhubs) == 0:
+            return []
 
-        # claim tasks from the taskhub
-        tasks = self.client.claim_taskhub_tasks(
-            taskhub, claimant=self.name, count=count
-        )
+        # claim tasks from taskhubs based on weight; keep going till we hit our
+        # total desired task count, or we run out of taskhubs to draw from
+        while len(tasks) < count and len(taskhubs) > 0:
+            # based on weights, choose taskhub to draw from
+            taskhub: List[ScopedKey] = random.choices(
+                list(taskhubs.keys()), weights=[tq.weight for tq in taskhubs.values()]
+            )[0]
+
+            # claim tasks from the taskhub
+            claimed_tasks = self.client.claim_taskhub_tasks(
+                taskhub,
+                compute_service_id=self.compute_service_id,
+                count=(count - len(tasks)),
+            )
+
+            # gather up claimed tasks, if present
+            for t in claimed_tasks:
+                if t is not None:
+                    tasks.append(t)
+
+            # remove this taskhub from the options available; repeat
+            taskhubs.pop(taskhub)
```

Comment:
Im not quite sure I understand this pop, what if we didn't take all the tasks from the hub? I'm probably missing something.

Reply:
The idea I've tried to implement here is explained in the comments. I'll try to express it in a different way:
1. When we enter this method, we want to claim `count` `Task`s. 
2. We start by randomly selecting a `TaskHub` to draw from based on their relative weights.
3. We try to draw `count` `Task`s from that `TaskHub`. However, it's possible that this `TaskHub` doesn't actually have enough `Task`s to meet our desired `count`. So, we drop it from the set we select from and roll the dice again, repeating from (2).
4. We keep going until we have either collected `count` `Task`s or run out of `TaskHub`s to draw `Task`s from.

Does the implementation above accomplish this, in your view?